src/kernels/gpu_reference_kernel/naive_conv.cpp

```diff
@@ -0,0 +1,1580 @@
+                                                    const float* __restrict__ p_wei,
+                                                    float* __restrict__ p_out,
+                                                    int hi,
+                                                    int wi,
+                                                    int n,
+                                                    int k_per_group,
+                                                    int c_per_group,
+                                                    int ho,
+                                                    int wo,
+                                                    int sy,
+                                                    int sx,
+                                                    int dy,
+                                                    int dx,
+                                                    int py,
+                                                    int px,
+                                                    int fy,
+                                                    int fx,
+                                                    int group)
+{
+    /*
+     *  need to compute total output pixel: `group * n * k_per_group * ho * wo`.
+     *  to distribute this workload, let one workgroup compute `ho * wo` pixel,
+     *  hence need `group * n * k_per_group` workgroups (grid_size).
+     */
+    int k             = k_per_group * group;
+    int c             = c_per_group * group;
+    int thread_length = ho * wo;
+    int bid           = blockIdx.x;
+    int ik            = bid % k_per_group;
+    int in            = (bid / k_per_group) % n;
+    int ig            = bid / (n * k_per_group);
+
+    p_in += in * c * hi * wi + ig * c_per_group * hi * wi;
+    p_wei += ig * k_per_group * c_per_group * fy * fx + ik * c_per_group * fy * fx;
+    p_out += in * k * ho * wo + ig * k_per_group * ho * wo + ik * ho * wo;
+
+    for(int tid = threadIdx.x; tid < thread_length; tid += 256)
+    {
+        int iho = tid / wo;
+        int iwo = tid % wo;
+
+        double value = .0f;
+
+        for(int ic = 0; ic < c_per_group; ic++)
+        {
+            for(int iy = 0; iy < fy; iy++)
+            {
+                int valid_h = 1;
+                int cur_h   = sy * iho - py + dy * iy;
+                if(cur_h < 0 || cur_h >= hi)
+                    valid_h &= 0;
+                for(int ix = 0; ix < fx; ix++)
+                {
+                    int valid_w = 1;
+                    int cur_w   = sx * iwo - px + dx * ix;
+                    if(cur_w < 0 || cur_w >= wi)
+                        valid_w &= 0;
+
+                    if(valid_w & valid_h)
+                    {
+                        int i_idx = ic * hi * wi + cur_h * wi + cur_w;
+                        int w_idx = ic * fy * fx + iy * fx + ix;
+                        value +=
+                            static_cast<double>(p_in[i_idx]) * static_cast<double>(p_wei[w_idx]);
```

Comment:
max `i_idx` is 2GiB, so max addressable size of float array is 8 GiB. Ditto `w_idx`.

Reply:
OK. I changed every address calculation to size_t

Reply:
After new commit to change address calculation with size_t, the above crash is fixed

Reply:
I afraid the fix is incorrect. Max `i_idx` is still 2GiB. Please let me know if you need more explanations of some C training.

Reply:
@atamazov sorry for the delay, I missed change the data type of like `i_idx`. This is fixed in latest commit